Restaurants with router

rheech22 opened the pull request:
트레이너님, 안녕하세요! 지금까지 작성한 내용 먼저 올리겠습니다 😃

rheech22 committed:
restaurants-router: Initial commit

rheech22 committed:
Feature: add react-router-dom

rheech22 committed:
Feature: add HomePage

rheech22 committed:
Feature: add AboutPage

rheech22 committed:
Feature: add Header

rheech22 committed:
Feature: add NotFoundPage

rheech22 committed:
Feature: Add RestaurantDetails Page and Container

rheech22 committed:
Refactor: RestaurantDetailsContainer with redux

rheech22 committed:
Feature: Attach Links on restaurant list

rheech22 committed:
Feature: click to see restaurant details

rheech22 committed:
Refactor: Edit a path for the detail page

moonkii commented on src/App.test.jsx:
```diff
@@ -2,33 +2,62 @@ import React from 'react';
+  }
+
+  beforeEach(() => {
+    const dispatch = jest.fn();
```
`dispatch` 를 외부에 선언해주시면 테스트에 필요할 때 사용하기 좋을거에요.

rheech22 replied:
넵 수정했습니다!

moonkii commented on src/RestaurantDetailsContainer.jsx:
```diff
@@ -0,0 +1,30 @@
+export default function RestaurantDetailsContainer() {
+  const restaurant = useSelector(get('restaurantDetails'));
+
+  if (!restaurant) {
+    return (
+      <h2>Loading...</h2>
+    );
+  }
```
이런 예외처리하는 디테일 좋아요! 👍

moonkii commented on src/RestaurantDetailsContainer.test.jsx:
```diff
@@ -0,0 +1,51 @@
+
+describe('RestaurantDetailsContainer', () => {
+  context('식당의 정보가 없다면', () => {
+    useSelector.mockImplementation((selector) => selector({
+      restaurantDetails: null,
+    }));
```
이렇게 매 테스트 마다 `useSelector` 를 반복해서 작성하면 불편하겠죠?
다른 테스트처럼 describe 내부에서 `beforeEach` 로 선언해주시고,
[given2](https://github.com/tatyshev/given2) 라는 라이브러리를 사용해서 lazy 하게 할당하는 방법을 사용해보세요!

rheech22 replied:
오 적용해보도록 하겠습니다!

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/133452274-a6f8ff2a-7fdc-4e92-bcc9-bf8d825c300f.png)

지금까지 잘 진행하고 있으신것 같아요! 👍 
TDD 사이클을 의식하면서 과정을 신경써서 진행해주세요 ;)

rheech22 committed:
Fix: change how it gets restaurant details

1. 후행 슬래시가 들어간 주소를 직접 입력하면 해당 주소에 접근하지 못하는 문제를 확인했습니다. index.html > script > 'main.js'를 '/main.js'로 수정하니 문제가 해결되었습니다.
2. 식당을 클릭하면 이벤트 핸들러를 통해 fetch하는 방식에서 디테일 페이지에서 useEffect로 주소의 params값을 얻어 fetch하는 방식으로 변경하였습니다.

rheech22 commented:
주소창에 id를 직접 입력하여 식당 정보를 불러올 수 있도록 리팩터링하는 과정에서 후행 슬래시가 들어간 주소는 route가 제대로 match시켜주지 못하는 문제를 확인했습니다.

rheech22 committed:
Test: RestaurantDetailPage component

rheech22 committed:
Test: fetchRestaurantDetails API

rheech22 committed:
Test: loadRestaurantDetails action

rheech22 committed:
Test: setRestaurantDetails reducer

rheech22 committed:
Fix: api parameters

rheech22 committed:
Refactor: test code with given2

rheech22 committed:
Refactor: test code with given2

rheech22 committed:
Refactor: test for reducer with given2

rheech22 committed:
Refactor: App test

moonkii commented:
> 검색을 해보다가 절대/상대 경로에 대해 알게 됐습니다. 아주 기초적인 내용 같은데 참 민망하네요 😅 앞에 슬래시를 붙이면 절대경로로 인식하기 때문에 하위 경로에서도 같은 조건으로 루트에 접근 가능한 것일까요?

라우팅을 하다보면 많이 겪는 문제인데 잘 해결하셨네요 ;)
나중에 개인프로젝트를 하시다보면 또 겪으실 수 있어서 잘 알아두시면 좋아요.
절대경로를 잡아주시면 아마 경로를 강제로 잡아주기 때문에 마찬가지로 하위에서도 접근이 가능할거에요.
main.js 는 어떤건지 혹시 알고계신가요?

moonkii commented:
given2 라이브러리를 잘 적용하신 것 같아요 👍

rheech22 commented:
> main.js 는 어떤건지 혹시 알고계신가요?

아뇨 😂   지나칠 수 있었던 부분을 짚어주셔서 감사합니다! 인터넷을 좀 참고해봤는데요. 웹팩으로 번들링된 결과가 main.js에 저장된 것이 맞나요??
